Propose an ADR about persistency (#640)
input-output-hk/hydra: docs/adr/2022-12-01-021-persistency.md

---
slug: 21
title: |
  21. Persistency
authors: []
tags: [Proposed]
---

## Status

Proposed

## Context

As any other process, a Hydra node can crash. If a head is open when it crashes then it's important that, when restarting the node, it can get to a meaningfull state regarding this head:
* who are the peers in the head?
* what was the last validated snapshot?
* what are the seen transaction?
* what is the status of the head (still open, closed, aborted)?

It's also important for an API client to be able to restore its own view of the world when re-connecting to a Hydra node after a restart.

All these persistent mechanisms should be reliable and have a limited performance impact on the system.

We've already implemented a form of state persistency in [#187](https://github.com/input-output-hk/hydra/issues/187)
and then brought some fixes in [#599](https://github.com/input-output-hk/hydra/issues/599)

We then have implemented a solution for an API client to recover its own state in [#580](https://github.com/input-output-hk/hydra/issues/580).

> ### Current solution

Reviewer: This is some nice documentation. Can we move it into the `core-concepts` section of the website and refer to it from here? That would make the ADR shorter and interested readers can learn about the current solution if the need (while others don't need that context always).

This is alos when we store them on disk that we give an id and a timestamp to each of these outputs:

```haskell
  appendToHistory history output = do
    time <- getCurrentTime
    timedOutput <- atomically $ do
      seq <- nextSequenceNumber <$> readTVar history
      let timedOutput = TimedServerOutput{output, time, seq}
      modifyTVar' history (timedOutput :)
      pure timedOutput
    append timedOutput
    pure timedOutput
```

The _append_ function uses [withBinaryFileDurable](https://hackage.haskell.org/package/unliftio-0.2.23.0/docs/UnliftIO-IO-File.html#v:withBinaryFileDurable)
 to append data on disk. Some reliability is ensured since we only append data to the file and never change existing data.
We may have problem if the last line of the file contains a partially writtent JSON object.

Also note that, with the current implementation, we open and close the file for every server-output we want to persist.

#### Drawbacks

The current solution has several drawbacks. We describe some of them here.

Extra disk usage with the state persistency:
* copy of the file before overriding everything;

> * Serialization of the whole state evrytime.

Reviewer: ```suggestion
* Serialization of the whole state everytime.
```

> The state is maintained in two several files, one for server-output and one for state which may lead to incoherences in some corner case situations.

Reviewer: ```suggestion
The state is maintained in two separate files, one for server-output and one for state which may lead to inconsistencies in some corner case situations.
```

When a customer send a _getUTxO_ query, for instance, the answer to this query is stored in the _server-out_ file and will then be replayed after a restart

> eventhough it wouldn't make much sense as the newly connected client didn't make any request. It could be helpful do differentiate between request answers

Reviewer: ```suggestion
even though it wouldn't make much sense as the newly connected client didn't make any request. It could be helpful do differentiate between request answers
```

and actual events which should be persisted.

When a new peer is seen, a message is sent to the API client and this is stored in _server-out_. So on restart, the message will be replayed after a restart

> and the client will be informed that the peer is connected eventhough it is certainly not yet connected, just after a restart. As a counter measure, we
> also send a _Greetings_ message, later, to notify the client that the hydra node has been restarted and that it should _forget_ about some information
> received before. But may be we could have a persistent that could distinguish volatile events from persistent ones?

Reviewer: ```suggestion
and the client will be informed that the peer is connected event hough it is certainly not yet connected, just after a restart. As a counter measure, we
also send a _Greetings_ message, later, to notify the client that the hydra node has been restarted and that it should _forget_ about some information
received before. But maybe we could distinguish volatile events from persistent ones?
```

### event/command sourcing as an alternative

[Event sourcing](https://martinfowler.com/eaaDev/EventSourcing.html) could be an alternative to store the state of the Hydra node. In that kind of architecture,
everytime something happens that could update the state, an event is generated, then stored on disk, then applied to the state.

> Some of the values could be:

Reviewer: ```suggestion
Some of the benefits:
```

> * More effecient perisitence of the state (only events are happended on disk, not the full state and no need for file copying);

Reviewer: ```suggestion
* More efficient persistence of the state (events are only appended on disk, not the full state and no need for file copying);
```

> * reliability as the state is only updated after the events have been persisted on disk.

Reviewer: This is tricky right? Persisting something to disk and being sure that it's definitely written is not easy (or impossible). But, I do agree it's _more_ reliable than writing an aggregated state. Especially if some appended events are missing, it's not as fatal.

> That may also allow us to improve client API replay of events by introducing some logic at the server API level to decide what is worth replaying and what is not.

Reviewer: Instead of outlining the potential benefits.. I would suggest you just do propose the decision and describe this as a consequence (and also more concrete please)

> It may be the case that it could also help us with hydra peers network communication issues: a peer could ask for the replay of a list of events from another peer?

Reviewer: I would recommend NOT to do this. Why should we trust a peer to say what happened? We _could_ use the multi-signed snapshots as a source of truth and fetch them from peers. But this sounds not like a low-hanging fruit / consequence of this ADR.

Author: My bad. I am not considering general peer to peer trust here but more of a way of replaying lost messages in communication. But maybe it's a bit far fetched and not super useful to mention here.

> It might be the case that not all the inputs of our system could be mapped to the concept of _command_ from event-sourcing architecture. In particulier changes

Reviewer: ```suggestion
It might be the case that not all the inputs of our system could be mapped to the concept of _command_ from event-sourcing architecture. In particular changes
```

observed on-chain. Here is a diagram of a possible _event sourcing_ inspired architecture for a Hydra node:

> ![](img/021-event-sourcing.png)

Reviewer: White is hard to read!

In the above diagram, all the outputs would have to be stored on disk before being propagated and integrated to the state.

An alternative would be to go with a command sourcing approach in which, instead of storing the events produced by our system
(that is the outputs of the head logic in the above diagram), we would store the commands received by the system (that is the
inputs of the head logic in the above diagram).

> As of this writing, it is not clear if it would be superior or more problematic to command sourcing instead of event sourcing.

Reviewer: I disagree with some of the labels you use. Receiving a message from the network of observing a Hydra protocol transaction on L1 are **events** to me. They are immutable facts, which we need to process & handle.

This is also true for state-changing commands (currently called `ClientInput`s). To the event-reducing system of a `hydra-node`, these are also events.

You once mentioned that CQRS could be a way to help disambiguate state-changing and non-state-changing _things_ and I do agree here. Especially on the client-facing side of the system, if we distinguish between commands and queries, we can clearly see that only commands may change the sate and need to be persisted.

So maybe.. this ADR should be about doing CQRS on client inputs (renaming them to `Command` and `Query` types), mapping only commands to events and consequently also persisting all events (to overcome some limitations outlined). WDYT?

Author: I would distinguish CQRS and event-sourcing, even though both work very well together.

I think Martin Fowler gives super clear explanation of both [event-sourcing](https://martinfowler.com/eaaDev/EventSourcing.html) and [CQRS](https://martinfowler.com/bliki/CQRS.html).

CQRS, Command Query Responsibility Segregation, is really about distinguishing the command channel from the query channel and the underlying models. This allow one to update data against one model, update that is eventually replicated to another model, more fitted for performant queries, for instance.

Event-sourcing is really about modeling the changes to a state as events.

Then, when we combine both, we can have :
* commands that reach the system
* apply the commands to a command model
* generate events which represent the changes the command would apply to the model
* from these events, we can make our command model evolve to be ready for the next command
* from these events, we can make any query model to evolve to be ready for the next queries

All that is to say that this command/event term is not mine but taken from CQRS and event-sourcing architecture so it has to be understood in this very context.

It's not yet clear to me what is an on-chain _event_ in this context. Also, it is not clear to me yet what are the different contexts. The DirectChain code might see on-chain _events_ as immutable facts but it might then decide to emit commands to the HeadLogic depending on this facts 🤷 

This needs to be super clear before making this ADR more than a draft I think.
